metaflow/plugins/cards/component_serializer.py

```diff
@@ -16,11 +23,368 @@ def get_card_class(card_type):
+        card_creator,
+        components=None,
+        logger=None,
+        no_warnings=False,
+        user_set_card_id=None,
+        runtime_card=False,
+        card_options=None,
+    ):
+        self._card_creator_args = dict(
+            card_uuid=card_uuid,
+            user_set_card_id=user_set_card_id,
+            runtime_card=runtime_card,
+            decorator_attributes=decorator_attributes,
+            card_options=card_options,
+            logger=logger,
+        )
+        self._card_creator = card_creator
+        self._last_layout_change = None
+        self._latest_user_data = None
+        self._last_refresh = 0
+        self._last_render = 0
+        self._render_seq = 0
+        self._logger = logger
+        self._no_warnings = no_warnings
+        self._warn_once = {
+            "update": {},
+            "not_implemented": {},
+        }
+        card_type = decorator_attributes["type"]
+
+        if components is None:
+            self._components = ComponentStore(
+                logger=self._logger,
+                card_type=card_type,
+                user_set_id=user_set_card_id,
+                components=None,
+            )
+        else:
+            self._components = ComponentStore(
+                logger=self._logger,
+                card_type=card_type,
+                user_set_id=user_set_card_id,
+                components=list(components),
+            )
+
+    def append(self, component, id=None):
+        self._components.append(component, id=id)
+
+    def extend(self, components):
+        self._components.extend(components)
+
+    def clear(self):
+        self._components.clear()
+
+    def _card_proc(self, mode):
+        self._card_creator.create(**self._card_creator_args, mode=mode)
+
+    def refresh(self, data=None, force=False):
+        self._latest_user_data = data
+        nu = time.time()
+
+        # FIXME: make `RUNTIME_CARD_MIN_REFRESH_INTERVAL` customizable at decorator level.
+        if nu - self._last_refresh < RUNTIME_CARD_MIN_REFRESH_INTERVAL:
+            # rate limit refreshes: silently ignore requests that
+            # happen too frequently
+            return
+        self._last_refresh = nu
+
+        # This block of code will render the card in `render_runtime` mode when:
+        # 1. refresh is called with `force=True`
+        # 2. Layout of the components in the card has changed. i.e. The actual elements in the component array have changed.
+        # 3. The last time the card was rendered was more the minimum interval after which they should be rendered.
+        last_rendered_before_minimum_interval = (
+            nu - self._last_refresh
+        ) > RUNTIME_CARD_RENDER_INTERVAL
+        layout_has_changed = (
+            self._last_layout_change != self.components.layout_last_changed_on
+            or self._last_layout_change is None
+        )
+
+        if force or last_rendered_before_minimum_interval or layout_has_changed:
```

Comment:
nit: can proably bypass stuff if force is true :)

Reply:
can you expand on that? I didn't quite get what meant

Reply:
I think I menat that the part before this can probably be skipped since we know we are going to force it.